### ISSUE TITLE
Implement `assume(b: bool);` intrinsic

### ISSUE THREAD

**Issue by CohenArthur** [Rust-GCC/gccrs#1561]
LLVM has the `llvm.assume` intrinsic, represented by `__builtin_assume` in clang. Since GCC does not have this intrinsic, we need to lower `core::intrinsics::assume(<condition>)` to the following pseudo code:

```c
if (!<condition>)
    __builtin_unreachable();
```

or
```c
__builtin_expect(<condition>, <true>);
```

which should hopefully enable us to gain some optimizations

**Comment by CohenArthur** [comment]
Re "assume" (if I got that correctly), does the following maybe help, very recent: <
https://gcc.gnu.org/g:08b51baddc53d64aa4c5e7a81ef3c4bf320293be
> "c++, c: Implement C++23 P1774R8 - Portable assumptions [PR106654]"? 

Thanks to @tschwinge

**Comment by tamaroning** [comment]
Hi, I would like to tackle on this.

**Comment by CohenArthur** [comment]
@tamaroning Great! Thank you :) Feel free to ask any questions you might have, I'd be more than happy to help

**Comment by tamaroning** [comment]
@CohenArthur 
Hi, I found the code below which handles intrinsic funtions.
https://github.com/Rust-GCC/gccrs/blob/71b8beb150d96548a2ebfb8ef964d14225e300f0/gcc/rust/backend/rust-compile-resolve-path.cc#L150-L156

This call to `commpile.compile (fntype)` returns a function declaration, but we want to compile `assume(b)` to `if(!b) __builtin_unreachable()`, which is not just a func call. So I think I need to change other part of source code to register `assume` intrinisic. Could you advise me?

**Comment by CohenArthur** [comment]
Hi @tamaroning sorry for not answering earlier! This is indeed a problem, which I'm also facing for other intrinsics. I think maybe creating an `#[inline_always]` function could solve the issue? But unsure. Otherwise we'll have to figure out a way to add these calls directly, but that's a big refactor.

**Comment by tamaroning** [comment]
@CohenArthur 
Thank you for answer.
> I think maybe creating an #[inline_always] function could solve the issue?

`#[inline(always)]` attribute can be added only to function decl with body, so I think it cannot be used with intrinisc functions.

**Comment by CohenArthur** [comment]
Yeah, I think this will need some major refactoring. I'll remove the `good-first-pr` label and we can think about other, easier intrinsics to work on. Math intrinsics are quite easy usually @tamaroning if you'd like to take a crack at some, I can give you a list.

**Comment by tamaroning** [comment]
@CohenArthur 
> Yeah, I think this will need some major refactoring.
Okay.

> if you'd like to take a crack at some, I can give you a list.

**Comment by CohenArthur** [comment]
@tamaroning this is the list of math intrinsics inside `gcc/rust/backend/rust-builtins.h`. It would be great if you could get some of them to work and add corresponding testcases. Their implementation should be similar to `sqrtf32` and `sinf32` that we already support.

```
//     sym::sqrtf32 => "llvm.sqrt.f32",
//     sym::sqrtf64 => "llvm.sqrt.f64",
//     sym::powif32 => "llvm.powi.f32",
//     sym::powif64 => "llvm.powi.f64",
//     sym::sinf32 => "llvm.sin.f32",
//     sym::sinf64 => "llvm.sin.f64",
//     sym::cosf32 => "llvm.cos.f32",
//     sym::cosf64 => "llvm.cos.f64",
//     sym::powf32 => "llvm.pow.f32",
//     sym::powf64 => "llvm.pow.f64",
//     sym::expf32 => "llvm.exp.f32",
//     sym::expf64 => "llvm.exp.f64",
//     sym::exp2f32 => "llvm.exp2.f32",
//     sym::exp2f64 => "llvm.exp2.f64",
//     sym::logf32 => "llvm.log.f32",
//     sym::logf64 => "llvm.log.f64",
//     sym::log10f32 => "llvm.log10.f32",
//     sym::log10f64 => "llvm.log10.f64",
//     sym::log2f32 => "llvm.log2.f32",
//     sym::log2f64 => "llvm.log2.f64",
//     sym::fmaf32 => "llvm.fma.f32",
//     sym::fmaf64 => "llvm.fma.f64",
//     sym::fabsf32 => "llvm.fabs.f32",
//     sym::fabsf64 => "llvm.fabs.f64",
//     sym::minnumf32 => "llvm.minnum.f32",
//     sym::minnumf64 => "llvm.minnum.f64",
//     sym::maxnumf32 => "llvm.maxnum.f32",
//     sym::maxnumf64 => "llvm.maxnum.f64",
//     sym::copysignf32 => "llvm.copysign.f32",
//     sym::copysignf64 => "llvm.copysign.f64",
//     sym::floorf32 => "llvm.floor.f32",
//     sym::floorf64 => "llvm.floor.f64",
//     sym::ceilf32 => "llvm.ceil.f32",
//     sym::ceilf64 => "llvm.ceil.f64",
//     sym::truncf32 => "llvm.trunc.f32",
//     sym::truncf64 => "llvm.trunc.f64",
//     sym::rintf32 => "llvm.rint.f32",
//     sym::rintf64 => "llvm.rint.f64",
//     sym::nearbyintf32 => "llvm.nearbyint.f32",
//     sym::nearbyintf64 => "llvm.nearbyint.f64",
//     sym::roundf32 => "llvm.round.f32",
//     sym::roundf64 => "llvm.round.f64",
```

**Comment by tamaroning** [comment]
@CohenArthur 
Thank you!
I will open a pull request after adding some of them.

**Comment by dkm** [comment]
A bit late, but `assume` and `expect` may share some mechanics, but they don't have the same semantics:
- the expression in `assume` is known to be true
- the expression in `expect` is expected to be evaluated as the 2nd argument, but not always (a quick look shows that it considers the expression as 10/90 instead of 50/50).
Not sure that lowering to something use expect would be correct.

**Comment by philberty** [comment]
this was added in c5c4897e8d677c239c1d1122a2c58038a26ec943
